front/lib/api/assistant/actions/retrieval.ts

```diff
@@ -1,33 +1,238 @@
 import { Authenticator } from "@app/lib/auth";
+import { Err, Ok, Result } from "@app/lib/result";
+import logger from "@app/logger/logger";
 import {
   DataSourceConfiguration,
-  DataSourceFilter,
+  isRetrievalConfiguration,
   RetrievalActionType,
+  RetrievalConfigurationType,
   RetrievalDocumentType,
   TimeFrame,
 } from "@app/types/assistant/actions/retrieval";
 import {
-  AgentActionInputsSpecification,
+  AgentActionSpecification,
   AgentConfigurationType,
 } from "@app/types/assistant/agent";
 import {
   AssistantAgentMessageType,
   AssistantConversationType,
+  AssistantUserMessageType,
 } from "@app/types/assistant/conversation";
 
+import { generateActionInputs } from "../agent";
+import { ModelMessageType } from "../conversation";
+
 /**
- * Inputs generation.
+ * TimeFrame parsing
  */
 
-export async function retrievalInputsSpecification(
-  auth: Authenticator,
-  configuration: AgentConfigurationType
-): Promise<AgentActionInputsSpecification> {
+// Attempts to parse a string representation of the time frame of the form `{k}{unit}` or `all`
+// where {k} is a number and {unit} is one of `d`, `w`, `m`, `y` for day, week, month, year.
+export function parseTimeFrame(raw: string): TimeFrame | null {
+  const r = raw.trim().toLowerCase();
+  if (r === "all") {
+    return null;
+  }
+
+  const m = r.match(/^(\d+)([dwmy])$/);
+  if (!m) {
+    return null;
+  }
+
+  const count = parseInt(m[1], 10);
+  if (isNaN(count)) {
+    return null;
+  }
+
+  let duration: TimeFrame["duration"];
+  switch (m[2]) {
+    case "d":
+      duration = "day";
+      break;
+    case "w":
+      duration = "week";
+      break;
+    case "m":
+      duration = "month";
+      break;
+    case "y":
+      duration = "year";
+      break;
+    default:
+      return null;
+  }
+
+  return {
+    count,
+    duration,
+  };
+}
+
+/**
+ * Model rendering of retrievals.
+ */
+
+export function renderRetrievalActionForModel(
+  action: RetrievalActionType
+): ModelMessageType {
+  let content = "";
+  for (const d of action.documents) {
+    let title = d.documentId;
+    for (const t of d.tags) {
+      if (t.startsWith("title:")) {
+        title = t.substring(6);
+        break;
+      }
+    }
+
+    let dataSourceName = d.dataSourceId;
+    if (d.dataSourceId.startsWith("managed-")) {
+      dataSourceName = d.dataSourceId.substring(8);
+    }
+
+    content += `TITLE: ${title} (data source: ${dataSourceName})\n`;
+    content += `REFERENCE: ${d.reference}\n`;
+
+    content += `EXTRACTS:\n`;
+    for (const c of d.chunks) {
+      content += `${c.text}\n`;
```

Comment:
in the current form, there's nothing that tells the model about the chunk boundary right ? Or is the chunk content already stripped of all `\n` ?

Reply:
Chunks don't have `\n` :+1: